src/page/slot.rs

```diff
@@ -334,16 +286,161 @@ impl<T, C: cfg::Config> Slot<T, C> {
+    /// `Slab` we want to insert a value into it, as the memory is not initialized
+    pub(super) fn initialize_state(&self) -> Option<Generation<C>> {
+        // Load the current lifecycle state.
+        let lifecycle = self.lifecycle.load(Ordering::Acquire);
+        let gen = LifecycleGen::from_packed(lifecycle).0;
+        let refs = RefCount::<C>::from_packed(lifecycle);
+
+        test_println!(
+            "-> initialize; state={:?}; gen={:?}; refs={:?}",
+            Lifecycle::<C>::from_packed(lifecycle),
+            gen,
+            refs
+        );
+
+        if refs.value != 0 {
+            test_println!("-> initialize while referenced! cancelling");
+            return None;
+        }
+
+        // Set the slot's state to NOT_REMOVED.
+        let new_lifecycle = gen.pack(Lifecycle::<C>::NOT_REMOVED.pack(0));
+        let actual = self
+            .lifecycle
+            .compare_and_swap(lifecycle, new_lifecycle, Ordering::AcqRel);
+        if actual != lifecycle {
+            // The slot was modified while we were inserting to it! It's no
+            // longer safe to insert a new value.
+            test_println!(
+                "-> modified during insert, cancelling! new={:#x}; expected={:#x}; actual={:#x};",
+                new_lifecycle,
+                lifecycle,
+                actual
+            );
+            return None;
+        }
+
+        Some(gen)
     }
+}
 
+// Slot impl which _needs_ an `Option` for self.item, this is for `Slab` to use.
+impl<T, C> Slot<Option<T>, C>
+where
+    C: cfg::Config,
+{
     fn is_empty(&self) -> bool {
         self.item.with(|item| unsafe { (*item).is_none() })
     }
+
+    /// Insert a value into a slot
+    ///
+    /// We first initialize the state and then insert the pased in value into the slot.
+    #[inline]
+    pub(super) fn insert(&self, value: &mut Option<T>) -> Option<Generation<C>> {
+        debug_assert!(self.is_empty(), "inserted into full slot");
+        debug_assert!(value.is_some(), "inserted twice");
+
+        let gen = self.initialize_state()?;
+        // Set the new value.
+        self.item.with_mut(|item| unsafe {
+            *item = value.take();
+        });
+
+        test_println!("-> inserted at {:?}", gen);
+
+        Some(gen)
+    }
+
+    /// Tries to remove the value in the slot
+    ///
+    /// This method tries to remove the value in the slot. If there are existing references, then
+    /// the slot is marked for removal and the next thread calling either this method or
+    /// `remove_value` will do the work instead.
+    #[inline]
+    pub(super) fn try_remove_value<F: FreeList<C>>(
+        &self,
+        gen: Generation<C>,
+        offset: usize,
+        free: &F,
+    ) -> Option<T> {
+        if self.mark_for_mutation(gen) {
+            None
+        } else {
+            // Otherwise, we can remove the slot now!
+            test_println!("-> remove deferred; can remove now");
```

Comment:
minor nit: i think this should have the actual name of the function:
```suggestion
            test_println!("-> try_remove_value; can remove now");
```